### ISSUE TITLE
Extend model serialization to composite models with a components needing specialised serialization

### ISSUE THREAD

**Issue by Sh4pe** [alan-turing-institute/MLJ.jl#678]
### Problem

When a `TunedModel` that contains a `XGBoostRegressor` is stored using `MLJ.save` and then restored using the `model(filename::String)` method, the deserialized version does not work. This is a short script to reproduce the problem:

```julia
using MLJ
using DataFrames

@load XGBoostRegressor

X = DataFrame(A=rand(10), B=rand(10))
y = rand(10)

original_machine = let
    base_model = XGBoostRegressor()
    tuned_model = TunedModel(
        model=base_model,
        tuning=Grid(),
        range=[range(base_model, :num_round, values=[9,10,11])],
    )
    m = machine(tuned_model, X, y)
    fit!(m, rows=1:10)
    m
end

# This prediction works
predicted_from_original = predict(original_machine, X[1:5, :])
@show predicted_from_original

deserialized_machine = let
    # careful here!
    f = "/tmp/x.machine.can_be_deleted"
    MLJ.save(f, original_machine)
    machine(f)
end

# Yet this prediction fails
predicted_from_deserialized = predict(deserialized_machine, X[1:5, :])
@show predicted_from_deserialized
```

The first prediction works, of course, but the second prediction fails with the stacktrace that I've stored in [this gist](https://gist.github.com/Sh4pe/b038a53eebf8e4c32fd0da1467ee7f5d). Something doesn't seem to be stored. If you run the same example but don't wrap the `XGBoostRegressor` in a `TunedMachine`, the deserialized version works.

### Diagnosis

My first diagnosis is that the [special `save` method](https://github.com/alan-turing-institute/MLJXGBoostInterface.jl/blob/96d928070e6829202a5354ab2a28c028662fa14c/src/MLJXGBoostInterface.jl#L670-L679) for the `XGBoostRegressor` is not called when the `TunedModel` is stored. So the aditional information that XGBoost stores is not loaded.

I haven't checked this, but this problem might also appear if you wrap any other model with a custom `save` method into a `TunedModel`.

### Versions

* MLJ.jl v0.14.1.
* MLJModels v0.12.7

**Comment by ablaom** [comment]
Thanks for reporting! 

You need MLJModels 0.12.7 for XGBoost serialization to work. Do you have it?

**Comment by Sh4pe** [comment]
Yes, I'm using MLJModels 0.12.7.

**Comment by Sh4pe** [comment]
Also, note that serialization of the plain XGBoostRegressor does work - see the first example in my script.

**Comment by ablaom** [comment]
Yes, serialization won't work for TunedModels if the model being wrapped has a specialised serialization method. Similarly, composite models created using @pipeline or @from_network won't work if one of the component models has a specialised serialization. 

Everything is of course fine (as far as I know) if all models are pure Julia.

Fixing these wrappers should be straigtforward: 

- [ ] `TunedModel`
- [ ] `EnsembleModel` 
- [ ] and `BinaryThresholdPredictor`

The general composite models may be more difficult: 

- [ ] serialization for composite models with "special" components

@Sh4pe In the mean time, you could use `EvoTrees` instead of `XGBoost` ??

**Comment by Sh4pe** [comment]
Hey @ablaom, thanks for suggesting the workaround. I can use `EvoTrees` and I found a workaround that is usable for me. But I'd still like `TunedModel`s to be serializable/deserializable.

**Comment by ablaom** [comment]
Transferred to https://github.com/JuliaAI/MLJSerialization.jl/issues/16

**Comment by ablaom** [comment]
@Sh4pe An update to let you know that MLJ serialisation has been redesigned and your issue is hopefully resolved in the MLJ 0.18 (MLJBase 0.20) release. Please open a new issue at MLJBase if you continue to have problems. 

[Applicable release notes](https://github.com/alan-turing-institute/MLJ.jl/releases/tag/v0.18.0)
[Manual entry on serialization](https://alan-turing-institute.github.io/MLJ.jl/dev/machines/#Saving-machines)
